src/modules/User/resolvers.ts

```diff
@@ -0,0 +1,70 @@
+import { queryField, mutationField, nonNull } from 'nexus/dist';
+import { compare, hash } from 'bcrypt';
+import { sign } from 'jsonwebtoken';
+import { AUTH_SIGNUP_ENABLED, JWT_EXPIRATION_PERIOD, JWT_SECRET } from '../../env';
+
+export const me = queryField('me', {
+  type: 'User',
+  resolve: (_, __, ctx) => {
+    if (!ctx.authVerificationResult || !('userId' in ctx.authVerificationResult)) {
+      throw new Error('Not authorized');
+    }
```

Comment:
Why do you need to separately check this and throw error inside the resolver if you have the whole `permissions` setup already? Because of the types? 
- If because of the types, can you please extend the types, so this check is not needed every time you access the `ctx.authVerificationResult` but the types are correctly propagated?
- If not because of the types, why there are need to be two places that check auth? Can we please use a single place inside `permissions`?

Reply:
- it's not possible to adjust the context via `shield`
- the jwt validity object arrives to the shield and gets declined if invalid.
- then jwt object gets passed to the actual function. 
- the previous action was not taken into account by the typing system.

I can only come up with a hack where `userId` of the context gets hardcoded to be the string at all times. By default (where the id is not known from the jwt) it's something like 'unknown';

Reply:
I'm not sure I understand your answers, since you didn't actually answer to my questions.

> it's not possible to adjust the context via shield

I didn't ask for that, nor do I expect this is needed

> the jwt validity object arrives to the shield and gets declined if invalid

This I didn't get. What is the validity object? What means `declined`?

> the previous action was not taken into account by the typing system

By previous, do you mean "jwt object gets passed to the actual function". Then, my first suggestion stands true: please extend `ctx` type to have correct type for `ctx.authVerificationResult.userId` or just expose it under `ctx.userId` if that's simpler

> it's something like 'unknown'

This sounds terrible, see my suggestion above to fix types, instead of coming up with terrible workarounds

Reply:
Please answer my initial questions here https://github.com/makerdao-ses/switchboard-boilerplate/pull/16#discussion_r1135625681

Reply:
> Why do you need to separately check this and throw error inside the resolver if you have the whole permissions setup already?

because of the types: ts cannot infer types because it does not see that the jwt verification has already happened.

So the following happens:
  1. data arrives to the endpoint
  2. the context is created: this implies calling the verification of the jwt token.
      - successful verification: user id is in the object `authVerificationResult`
      - unsuccessful verification: error identifier is in the object `authVerificationResult`
      - no token: null instead of the object
  3. If it's a shielded endpoint, the shielding logic activates and checks that the object formed on stage (2) is not indicating any errors
  4. endpoint logic is activated.

Types are generated by nexus, then in the respective `me` resolver ts expects the context as it has initially been defined. Regardless what happens on the stages (2) and (3). It's not a transfrom pipeline (`.validateToken().getMe()`) but a collection of building blocks for which typing has been generated (`makeApi ( tokenVerifier, getMeResolver )`) - so since each building block is not directly tied to another in the codebase (instead just accepts common context) , it cannot be known via inference what is the verification result.

I do not see a way where it's possible to make ts believe that the verification has happened and the user id is present for sure.

Reply:
> Why can't you type the context to have correct union type to include all those types no matter if any of those building blocks executed or not?

this is (was) the case. and on successful verification it's not possible for ts to know if the id is present in the object or not because it expects the context as it is defined (your 3 bullet points).

> The other common way is to provide function into the context:

implemented this